chore: publishing config migration cleanup

hikchoi opened the pull request:
# chore: publishing config migration cleanup

This PR
- fixes migration and its tests' version to be `0.83.0`, which is the actual version that it will be introduced for general release.
- exposes latest migration in `RunMigrationCommand` quickpick (was omitted)
- updates dendron.yml for test-workspace
- deprecates `publishing.previewPort`, `publishing.siteRepoDir`, and `publishing.hierarchy.*.noIndexByDefault` 

## Code

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)



## Instrumentation

### Basics

- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 



## Tests

### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [~] Common cases tested
- [~] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)

- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome



## Docs

### Basics

- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 


### Basics

- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

## 



## Close the Loop

### Basics

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

hikchoi commented:
note: 
- this is branched off of https://github.com/dendronhq/dendron/tree/chore/add-json-validation-for-v5-config
- I will rebase to master once https://github.com/dendronhq/dendron/pull/2411 goes in
- doc update pending

hikchoi commented:
converting this to draft as config migration isn't actually going out with v82. I'll clean it up again and re-open.

hikchoi commented:
docs: https://github.com/dendronhq/dendron-site/pull/406

kevinslin reviewed (commented):
can we convert this to regular pr now? this should come out in 0.83

hikchoi commented on packages/common-all/src/constants/configs/compat.ts:
```diff
@@ -3,5 +3,5 @@ export const CONFIG_TO_MINIMUM_COMPAT_MAPPING: { [key: number]: string } = {
   3: "0.65.0", // config consolidation (workspace namespace)
   4: "0.70.0", // config consolidation (preview namespace)
-  5: "0.81.0", // config consolidation (publishing namespace) TODO: FIX THIS BEFORE GENERAL RELEASE
+  // 5: "0.83.0", // config consolidation (publishing namespace) TODO: FIX THIS BEFORE GENERAL RELEASE
```
I've commented this out because adding this compat mapping would prevent users from keeping the v4 config and still use the cli for publishing.

If we want to force the migration I can re-enable it.

kevinslin replied:
looks good, can you also add this comment in the code?

kevinslin commented:
write now, how do we manually test this? when I run this from the debugger, it doesn't migrate the configuration (presumably because the current release version is 0.82.0)

kevinslin reviewed (commented):
in the modal dialogue where we tell users that this will change publishing, can we also have a link in the modal that explains what they need to run? we can add some docs and a link to [[Migration|dendron://dendron.dendron-site/dendron.topic.publish.migration]]

hikchoi commented:
> how do we manually test this? when I run this from the debugger, it doesn't migrate the configuration (presumably because the current release version is 0.82.0)

I tested with the current version temporarily hardcoded as 0.83.0 where the migration logic reads the current version

[here](https://github.com/dendronhq/dendron/blob/chore/publishing-namespace-migration-cleanup/packages/plugin-core/src/_extension.ts#L528:L528) if you want to test it migrates when you click on the toast message button

kevinslin commented:
> 

heads up that I wrote upgrade instructions from 11ty at [[Migration|dendron://dendron.dendron-site/dendron.topic.publish.migration]]. can you link to it from the modal?

hikchoi commented:
> > 
> 
> heads up that I wrote upgrade instructions from 11ty at [[Migration|dendron://dendron.dendron-site/dendron.topic.publish.migration]]. can you link to it from the modal?

I was just reading through the changes :) I'll add it to the modal

hikchoi commented:
update:

- modal now links to `[[Migration|dendron://dendron.dendron-site/dendron.topic.publish.migration]]`
  - updated the docs here: https://github.com/dendronhq/dendron-site/pull/406/files#diff-1ba81a8459e10d85c4fcd0ee344746041d531f2eea7af49b4e3571b2f9c9972dR9
- `generateChangelog` and `enableContainers` now deprecated and removed from migration logic

jonathanyeung commented:
merging into release 83 branch + master
